synapse/handlers/federation_event.py

```diff
@@ -766,15 +766,22 @@ async def _process_pulled_event(
         """
         logger.info("Processing pulled event %s", event)
 
-        # TODO: Why does this matter? The whole point of this function is to
-        # persist random PDU's from backfill. It shouldn't matter whether we saw
-        # them somewhere else first as an outlier, then during backfill. This
-        # function handles de-outliering anyway.
+        # This function should only be used to backfill events. If you're trying
+        # to persist an outlier, use another method. If you happen to run into a
+        # situation where the event you're trying to backfill is marked as an
+        # `outlier`, then you should update that spot to return an `EventBase`
+        # without the `outlier` flag set.
         #
-        # these should not be outliers.
-        assert (
-            not event.internal_metadata.is_outlier()
-        ), "pulled event unexpectedly flagged as outlier"
+        # `EventBase` is used to represent both an event we have not yet
+        # persisted, and one that we have persisted and now keep in the cache.
+        # In an ideal world this method would only be called with the first type
+        # of event, but it turns out that's not actually the case and for
+        # example, you could get an event from cache that is marked as an
+        # `outlier` (fix up that spot though).
+        assert not event.internal_metadata.is_outlier(), (
+            "This is a safe-guard to make sure you're not trying to persist an outlier using this function (use something else). "
+            "If you're trying to backfill an event, this is the right method but you need pass in an event copy that doesn't have `event.internal_metada.outlier = true`."
```

Comment:
bearing in mind this is going to pop up in a user's logs if/when we mess up, I'm not sure this level of detail is that helpful. By all means include these details in a comment, but I'd keep the assertion message short and sweet ("outlier event passed to `_process_pulled_event`" or something)

Reply:
Comment moved to the other PR: https://github.com/matrix-org/synapse/pull/13320#discussion_r923903077